ouroboros-consensus-byron/src/Ouroboros/Consensus/Byron/Ledger/NetworkProtocolVersion.hs

```diff
@@ -31,9 +31,12 @@ instance HasNetworkProtocolVersion ByronBlock where
 
 instance SupportedNetworkProtocolVersion ByronBlock where
   supportedNodeToNodeVersions   _ = Map.fromList [
-        (NodeToNodeV_1, ByronNodeToNodeVersion1)
+        (NodeToNodeV_1, ByronNodeToNodeVersion1),
         -- V_2 enables block size hints for Byron headers within the hard fork
         -- combinator, not supported by Byron-only.
+        (NodeToNodeV_2, ByronNodeToNodeVersion1),
+        (NodeToNodeV_3, ByronNodeToNodeVersion1),
+        (NodeToNodeV_4, ByronNodeToNodeVersion1)
       ]
```

Comment:
I think adding V_2, V_3, V_4 is wrong.

The months before and after the time the hard fork happened, Byron-only nodes were deployed running V_1. The new Cardano node would have to be backwards-compatible with the deployed Byron-only nodes (up to the start of the Shelley era, after which they'd disconnect). This means that Cardano nodes running V_1 were binary compatible with Byron-only nodes running V_1. Cardano nodes also had a V_2, which includes the era tag (and thus support for Shelley), which was not binary compatible with Byron-only nodes. Byron-only nodes intentionally *did not have V_2*, because they don't know about the era tag. So Byron-Byron and Byron-Cardano communication was done using V_1 using the Byron-only encoding, and Cardano-Cardano communication was done using V_2 using the Cardano (era wrapper) encoding.

So if you now add V_2 to Byron-only nodes, using the same Byron-only encoding, Byron-Cardano communication will use V_2, but with different encodings :boom:. Same for V_3. This means backwards compatibility between Byron and Cardano is broken.

However, in practice, no new Byron-only nodes will be deployed, so this won't cause real problems.

Still, I think that it would be morally wrong to add those new versions, because it breaks the backwards compatibility that we built in for no good reason.

Moreover, there is no reason to add them here since we won't be running Byron-only nodes.

Reply:
We don't want to hold `V_1` forever.  There are no _byron-only_ nodes any more with which we need to be compatible.   Do you have plans to drop support for _byron-only_ nodes in the future? That would be another way for dropping `V_1` for us.

Reply:
> Do you have plans to drop support for *byron-only* nodes in the future?

That's determined by the `cardano-node`, not by consensus. @dcoutts Any plans for that?

I'm okay with dropping `V_1` since we are indeed not going to deploy new Byron-only nodes that have to be compatible with Cardano nodes.

I propose we do the same thing as I suggested for Shelley-only: only list `V_4` and drop the older versions.

It would be good to include a comment saying that we used to only support V_1 for the reasons I mentioned (mainly so that the backwards compat story told in the Cardano encoder/decoders code still makes sense). I would be happy to push a commit adding such a comment.

Reply:
I included `NodeToNodeV_4` only with a comment.

Reply:
It's unclear if we will drop support for a Byron-only mode, however we certainly do not need network binary compatibility with those earlier nodes. So if someone for some strange reason wants to run a Byron-only testnet, it would be totally fine to do that using a more recent supported network protocol version.

So to a first approximation, we no longer have any binary compat constraints for Byron-only nodes, because there are no such deployed nodes, only hypothetical testnets for the curious code archeologists or the insane.

Reply:
> or the insane
:D